plugins/search-backend-node/src/indexing/BatchSearchEngineIndexer.ts

```diff
@@ -110,6 +110,11 @@ export abstract class BatchSearchEngineIndexer extends Writable {
    * @internal
    */
   async _final(done: (error?: Error | null) => void) {
+    const maybeError = await this.initialized;
```

Comment:
If this throws, it'll still never call `done`. Could this instead be done inside the try/catch?

Reply:
The failure is already caught on line 51->53. It took me a while to get my head around the promise created on line 44.

Admittedly, my javascript is not as good as others, but I think the following might be equivalent and easier to understand, what do you think:

```
  constructor(options: BatchSearchEngineOptions) {
    super({ objectMode: true });
    this.batchSize = options.batchSize;
    this.initialized = this.initialize();
  }

...
    try {
      await this.initialized;
    } catch (maybeError) {
      done(maybeError);
      return;
    }
```

Reply:
Am I missing something that justifies using setImmediate?

Reply:
Yeah like the comment alludes to, it ensures that the execution of the actual work happens one Node tick later. That way, when the constructor returns, the initialize call hasn't even been started yet, and won't happen until the current pool of pending work is exhausted (I think that's the correct wording ... but I would need to double check). Since this is a base class, child classes will call `super()` and then do MORE work such as setting member variables, that might need to be in place for `initialize` to work correctly. Thus - init happens one tick later which is guaranteed to be after the child constructor returns.

Reply:
You are correct in saying that it's actually safe now. But just to be sure that a future refactor doesn't _introduce_ the ability for the promise to throw and silently re-introduce the problem ... do move the exact code you already have, unchanged into the `try` block. Then I think we can ship it!

Reply:
Ah right ok. Sounds like setImmediate is required then. I have moved the wait into the try block.